src/placement.c

```diff
@@ -0,0 +1,548 @@
+// SPDX-License-Identifier: GPL-2.0-only
+#include <assert.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include "common/macros.h"
+#include "labwc.h"
+#include "placement.h"
+#include "ssd.h"
+#include "view.h"
+
+#define overlap_bitmap_index(bmp, i, j) \
+	(bmp)->grid[i * ((bmp)->nr_cols - 1) + j]
+
+struct overlap_bitmap {
+	int nr_rows;
+	int nr_cols;
+	int *rows;
+	int *cols;
+	int *grid;
+};
+
+static int
+compare_ints(const void *a, const void *b)
+{
+	return *(const int *)a - *(const int *)b;
+}
+
+static void
+destroy_bitmap(struct overlap_bitmap *bmp)
+{
+	assert(bmp);
+
+	if (bmp->rows) {
+		free(bmp->rows);
+	}
+
+	bmp->nr_rows = 0;
+	bmp->nr_cols = 0;
+	bmp->rows = NULL;
+	bmp->cols = NULL;
+	bmp->grid = NULL;
+}
+
+/* Count the number of views on view->output, excluding *view itself */
+static int
+count_views(struct view *view)
+{
+	assert(view);
+
+	struct server *server = view->server;
+	assert(server);
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		return 0;
+	}
+
+	int nviews = 0;
+
+	struct view *v;
+	for_each_view(v, &server->views, LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
+		/* Ignore the target view or anything on a different output */
+		if (v == view || v->output != output) {
+			continue;
+		}
+
+		nviews++;
+	}
+
+	return nviews;
+}
+
+/* Sort and de-deplicate a list of points that define a 1-D grid */
+static int
+order_grid(int *edges, int nedges)
+{
+	/* Sort grid edges */
+	qsort(edges, nedges, sizeof(int), compare_ints);
+
+	/* Skip over non-unique edges, counting the unique ones */
+	/* This is taken almost verbatim from Openbox. */
+	int i = 0;
+	int j = 0;
+
+	while (j < nedges) {
+		int last = edges[j++];
+		edges[i++] = last;
+		while (j < nedges && edges[j] == last) {
+			++j;
+		}
+	}
+
+	return i;
+}
+
+/*
+ * Construct an irregular grid that divides the usable area of view->output
+ * by extending the edges of every view on the output (except for *view itself)
+ * to infinity. The resulting grid will consist of rectangular intervals that
+ * are either completely uncovered by any view, or entirely covered.
+ * Furthermore, when any view intersects any interval on the grid, that view
+ * overlaps the whole interval: no view ever partially intersects any interval.
+ */
+static void
+build_grid(struct overlap_bitmap *bmp, struct view *view)
+{
+	assert(bmp);
+	assert(view);
+
+	struct server *server = view->server;
+	assert(server);
+
+	/* Always start with a fresh bitmap */
+	destroy_bitmap(bmp);
+
+	struct output *output = view->output;
+	if (!output_is_usable(output)) {
+		return;
+	}
+
+	int nviews = count_views(view);
+	if (nviews < 1) {
+		return;
+	}
+
+	/* Number of rows/columns is bounded by two per view plus screen edges */
+	int max_rc = 2 * nviews + 2;
+
+	int rows[max_rc];
```

Comment:
If you agree with the comment below about doing separate allocations for row/col/grid, I think it would seem cleaner to use `wl_array` here (which is just a simple array implementation). It would save us having to first create `rows[]` and `cols[]` here only to then heap allocate the same.

I haven't tried it, but think `qsort()` works on a `wl_array` array.

Reply:
I moved from `malloc` to `xzalloc` as recommended and avoided the stack allocation by just directly allocating (once for each of the arrays, rather than folding them all into a single allocation) enough storage on the heap to store the maximum grid size right in the bitmap struct. I don't think `wl_array` buys anything here except extra complexity.

For one, it will tend to over-allocate on the heap even worse than just asking for the maximum possible grid size up front; `wl_array_add` looks for power-of-two sizes to satisfy the requested addition. For another, we will still be directly manipuatling the underlying data array everywhere, so this would just replace a single call to `xzalloc` with a structure delcaration, `wl_array_init` call and `wl_array_add` call; and all subsequent data access currently of the form `bmp->member` with `bmp->member->data`. This also ties the implementation of the placement algorithm, where the array manipulations are entirely contained, to the that of `wl_array` and might require a refactor of the placement algorithms should the implementation of `wl_array` ever change.

If you really want to see these moved to `wl_array`, I will do so; but I'd rather just leave them as low-level C arrays.

Reply:
Good points. Normal C arrays are fine.